From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrew Branson <andrew.branson@jolla.com>
Date: Wed, 5 Aug 2026 17:53:32 +0200
Subject: Preload autoplay metadata before inaudible check

Load metadata for blocked autoplay elements so Gecko can discover that a
video has no audio track, then recompute preload once metadata is known.
---
 dom/media/mediaelement/HTMLMediaElement.cpp | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dom/media/mediaelement/HTMLMediaElement.cpp b/dom/media/mediaelement/HTMLMediaElement.cpp
index aa7f010579c8..438295c8a84e 100644
--- a/dom/media/mediaelement/HTMLMediaElement.cpp
+++ b/dom/media/mediaelement/HTMLMediaElement.cpp
@@ -3294,6 +3294,8 @@ void HTMLMediaElement::UpdatePreloadAction(
   // as we'll need it to play.
   if ((AllowedToPlay() && HasAttr(nsGkAtoms::autoplay)) || !mPaused) {
     nextAction = HTMLMediaElement::PRELOAD_ENOUGH;
+  } else if (HasAttr(kNameSpaceID_None, nsGkAtoms::autoplay)) {
+    nextAction = HTMLMediaElement::PRELOAD_METADATA;
   } else {
     // Find the appropriate preload action by looking at the attribute.
     const nsAttrValue* val =
@@ -6152,6 +6154,9 @@ void HTMLMediaElement::MetadataLoaded(const MediaInfo* aInfo,
   mTags = std::move(aTags);
   mLoadedDataFired = false;
   ChangeReadyState(HAVE_METADATA);
+  if (HasAttr(kNameSpaceID_None, nsGkAtoms::autoplay)) {
+    UpdatePreloadAction(JSCallingLocation::Get());
+  }
 
   // Add output tracks synchronously now to be sure they're available in
   // "loadedmetadata" event handlers.
